4.7 integration basics updates (#112)
smartdevicelink/sdl_android_guides: docs/Getting Started/Integration Basics/index.md

## SmartDeviceLink Service

> A SmartDeviceLink Android Service should be created to manage the lifecycle of the SdlManager. The SDL Service enables auto-start by creating the SdlManager, which then waits for a connection from SDL. This file also sends and receives messages to and from SDL after connected.

Reviewer: ```suggestion
A SmartDeviceLink Android Service should be created to manage the lifecycle of the SDL session. The `SdlServierce` should build and start an instance of the `SdlManager` which will automatically connect with a headunit when available. This `SdlManager` will handle sending and receiving messages to and from SDL after connected.
```

}
```

### Implementing SDL Manager

> In order to correctly connect to an SDL enabled head unit developers need to implement methods for the proper creation and disposing of an SdlManager in our SDLService.

Reviewer: ```suggestion
In order to correctly connect to an SDL enabled head unit developers need to implement methods for the proper creation and disposing of an `SdlManager` in our `SdlService`.
```

!!! NOTE
An instance of SdlManager cannot be reused after it is closed and properly disposed of. Instead, a new instance must be created. Only one instance of SdlManager should be in use at any given time.
!!!

```java
public class SdlService extends Service {

    //The manager handles communication between the application and SDL
    private SdlManager sdlManager = null;

    //...

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        
        if (sdlManager == null) {
            MultiplexTransportConfig transport = new MultiplexTransportConfig(this, APP_ID, MultiplexTransportConfig.FLAG_MULTI_SECURITY_OFF);
           
            // The app type to be used
            Vector<AppHMIType> appType = new Vector<>();
            appType.add(AppHMIType.MEDIA);

            // The manager listener helps you know when certain events that pertain to the SDL Manager happen
            SdlManagerListener listener = new SdlManagerListener() {
                
                @Override
                public void onStart() {

>                 	// RPC listeners and other functionality can be called once this callback is triggered.

Reviewer: ```suggestion
                	// After this callback is triggered the SdlManager can be used to interact with the conencted SDL session (updating the display, sending RPCs, etc)
```

                }

                @Override
                public void onDestroy() {
                    SdlService.this.stopSelf();
                }

                @Override
                public void onError(String info, Exception e) {
                }
            };

            // Create App Icon, this is set in the SdlManager builder
            SdlArtwork appIcon = new SdlArtwork(ICON_FILENAME, FileType.GRAPHIC_PNG, R.mipmap.ic_launcher, true);

            // The manager builder sets options for your session
            SdlManager.Builder builder = new SdlManager.Builder(this, APP_ID, APP_NAME, listener);
            builder.setAppTypes(appType);
            builder.setTransportType(transport);
            builder.setAppIcon(appIcon);
            sdlManager = builder.build();
            sdlManager.start();
        }

}
```

> `onDestroy()` is called whenever the manager detects some disconnect in the connection, whether initiated by the app, by SDL, or by the device’s connection.

Reviewer: ```suggestion
The `onDestroy()` method from the `SdlManagerListener` is called whenever the manager detects some disconnect in the connection, whether initiated by the app, by SDL, or by the device’s connection.
```

!!! IMPORTANT

> We must properly stop the service in the `onDestroy()` method as shown above.

Reviewer: ```suggestion
The `sdlManager` must be shutdown properly in the `SdlService.onDestroy()` callback using the method `sdlManager.dispose()`.
```

!!!

> ### Listening for events

Reviewer: ```suggestion
### Listening for RPC notifications and events
```

> We can listen for specific events using `SdlManager`'s `addOnRPCNotificationListener`. These listeners can be placed in the `onStart()` callback of the `SdlManagerListener`. We give an example of using this below, to obtain the HMI Status. Other listeners for specific RPCs can be added easily by changing the `FunctionID` to that of the RPC that is to be listened for and by casting the notification to the appropriate type.

Reviewer: ```suggestion
We can listen for specific events using `SdlManager`'s `addOnRPCNotificationListener`. These listeners can be added either in the `onStart()` callback of the `SdlManagerListener` or after it has been triggered. The following example shows how to listen for HMI Status notifications. Additional listeners can be added for specific RPCs by using their corresponding `FunctionID` in place of the `ON_HMI_STATUS` in the following example. 
```

Make sure this local class (SdlRouterService.java) is in the same package of SdlReceiver.java (described below)
!!!

If you created the service using the Android Studio template then the service should have been added to your `AndroidManifest.xml` otherwise the service needs to be added in the manifest. Because we want our service to be seen by other SDL enabled apps, we need to set `android:exported="true"`. The system may issue a lint warning because of this, so we can suppress that using `tools:ignore="ExportedService"`.  

> We also need to define the activity used for the lock screen. For more information about lock screens, please see the [Adding the Lock Screen](https://smartdevicelink.com/en/guides/android/adding-the-lock-screen/) section.

Reviewer: ```suggestion
## Lock Screen Activity 

An Activity entry must also be added to the manifest for the SDL lockscreen. For more information about lock screens, please see the [Adding the Lock Screen](https://smartdevicelink.com/en/guides/android/adding-the-lock-screen/) section.
```
